compiler/src/dotty/tools/dotc/transform/LazyVals.scala

```diff
@@ -274,168 +282,212 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
-    *       return value_0
-    *     } else if (state == <state-0>) {
-    *       if (LazyVals.CAS(this, bitmap_offset, flag, <state-1>, <field-id>)) {
-    *         try {
-    *           val result = <RHS>
-    *           value_0 = result
-    *           nullable = null
-    *           LazyVals.setFlag(this, bitmap_offset, <state-3>, <field-id>)
-    *           return result
-    *         }
-    *         catch {
-    *           case ex =>
-    *             LazyVals.setFlag(this, bitmap_offset, <state-0>, <field-id>)
-    *             throw ex
-    *         }
-    *       }
-    *     } else /* if (state == <state-1> || state == <state-2>) */ {
-    *       LazyVals.wait4Notification(this, bitmap_offset, flag, <field-id>)
-    *     }
-    *   }
-    * }
-    * ```
-    */
+  /**
+   * Create a threadsafe lazy accessor equivalent to the following code:
+   * ```
+   * private @volatile var _x: AnyRef = null
+   * @tailrec def x: A =
+   *    _x match
+   *    case current: A =>
+   *        current
+   *    case null =>
+   *        if CAS(_x, null, Evaluating) then
+   *            var result: AnyRef = null // here, we need `AnyRef` to possibly assign `NULL`
+   *            try
+   *                result = rhs
+   *                nullable = null // if the field is nullable; see `CollectNullableFields`
+   *                if result == null then result = NULL // drop if A is non-nullable
+   *            finally
+   *                if !CAS(_x, Evaluating, result) then
+   *                    val lock = _x.asInstanceOf[Waiting]
+   *                    CAS(_x, lock, result)
+   *                    lock.release()
+   *        x
+   *    case Evaluating =>
+   *        CAS(_x, Evaluating, new Waiting)
+   *        x
+   *    case current: Waiting =>
+   *        current.awaitRelease()
+   *        x
+   *    case NULL => null
+   * ```
+   * Where `Evaluating` and `NULL` are represented by `object`s and `Waiting` by a class that
+   * allows awaiting the completion of the evaluation. Note that since tail-recursive
+   * functions are transformed *before* lazy-vals, this implementation directly implements
+   * the resulting loop. `PatternMatcher` coming before `LazyVals`, the pattern matching block
+   * is implemented using if-s. That is:
+   * 
+   * ```
+   * private @volatile var _x: AnyRef = null
+   * def x: A =
+   *   while true do
+   *     val current: AnyRef = _x
+   *     if current == null then
+   *       if CAS(_x, null, Evaluating) then
+   *         var result: AnyRef = null
+   *         try
+   *           result = rhs
+   *           nullable = null
+   *           if result == null then result = NULL
+   *         finally
+   *           if !CAS(_x, Evaluating, result) then
+   *             val lock = _x.asInstanceOf[Waiting]
+   *             CAS(_x, lock, result)
+   *             lock.release()
+   *     else
+   *       if current.isInstanceOf[A] then
+   *         current.asInstanceOf[A]
```

Comment:
This must come last. If `A` erases to `Object`, this test will be `true` for any non-`null` value, *including* instances of `NULL`, `Waiting` and `Evaluating`!